File: python/src/dashgps/formats/ligo_ts_trailer.py
"""LigoGPS trailer appended after the last TS packet.

Implements spec/01-ligogps-ts-trailer.md. Clause numbers in comments refer to §1.6.

Status: VERIFIED against ~1,230 real ICESKY files. This is the format dashgps exists for.
"""

from ..io import read_slab
from ..ligo_record import (
    LIGO_MAGIC,
    detect_stride,
    is_ascii_region,
    is_date_shape,
    parse_ligo_text,
    record_text,
)
from ..model import ParseError, ParseResult, Point

FORMAT_ID = "ligo.ts_trailer"
STATUS = "verified"

END_MAGICS = (b"####", b"&&&&")
TAGS = (b"SKIP", b"skip")
FOOTER_PROBE = 65536  # the tail sniff budget, spec 00. One read covers a typical
                      # whole trailer, so a clip costs a single seek.
MAX_BLOCKS = 256


def _memo(opts, value):
    if opts is not None:
        opts.probe["ligo.footer"] = value
    return value


def _footer(reader, opts=None):
    """Clause 1. Returns (slab, trailer_len, magic) or None.

    Memoized on opts.probe: sniff and parse both need the footer, and on a network filesystem a
    second round trip per file is the difference between seconds and minutes over a big folder.
    """
    if opts is not None and "ligo.footer" in opts.probe:
        return opts.probe["ligo.footer"]
    n = reader.size()
    if n < 32:
        return _memo(opts, None)
    slab = read_slab(reader, n - FOOTER_PROBE if n > FOOTER_PROBE else 0, n)
    if not slab.covers(n - 8, 8):
        return _memo(opts, None)
    magic = slab.bytes(n - 8, 4)
    if magic not in END_MAGICS:
        return _memo(opts, None)
    length = slab.u32be(n - 4)
    if length < 8 or length > n:
        return _memo(opts, None)
    out = (slab, length, magic)
    if opts is not None:
        opts.probe["ligo.footer"] = out
    return out


def sniff(reader, opts):
    return 0.95 if _footer(reader, opts) is not None else 0.0


def _read_header(slab, off, tr_end):
    """Block header at off, or None. Clause 3."""
    if off + 8 > tr_end or not slab.covers(off, 8):
        return None
    if slab.bytes(off + 4, 4) not in TAGS:
        return None
    return slab.u32be(off)


def _resync(slab, after, tr_end, tr_start, res):
    """Find the next plausible block header. Clause 3."""
    p = after
    while True:
        a = slab.find(b"SKIP", p, tr_end)
        b = slab.find(b"skip", p, tr_end)
        if a < 0:
            hit = b
        elif b < 0:
            hit = a
        else:
            hit = a if a < b else b
        if hit < 0:
            return -1
        start = hit - 4
        if start >= slab.base and _read_header(slab, start, tr_end) is not None:
            sz = slab.u32be(start)
            if 8 <= sz <= (tr_end - start) + 8:
                # Offset from the start of the trailer: stable regardless of how much
                # of the file the reader happened to buffer.
                res.warn("resynchronised to block header at trailer+%d" % (start - tr_start))
                return start
        p = hit + 4


def parse(reader, opts):
    res = ParseResult(FORMAT_ID, STATUS, sources=[reader.name])
    n = reader.size()

    f = _footer(reader, opts)
    if f is None:
        raise ParseError("no LigoGPS trailer footer")
    foot, length, magic = f
    res.meta["end_magic"] = magic.decode("ascii")
    res.meta["trailer_len"] = length

    # Clause 2. Reuse the footer probe when the whole trailer already fits inside it.
    tr_start = n - length
    read_from = tr_start
    if length > opts.tail_cap:
        read_from = n - opts.tail_cap
        res.warn(
            "trailer is %d bytes, larger than the %d-byte cap; parsing the tail only"
            % (length, opts.tail_cap)
        )
    slab = foot if foot.base <= read_from else read_slab(reader, read_from, n)
    tr_end = n - 8

    # Clause 3: walk the blocks.
    blocks = []
    pos = tr_start if tr_start >= slab.base else slab.base
    guard = 0
    while pos + 8 <= tr_end and guard < MAX_BLOCKS:
        guard += 1
        sz = _read_header(slab, pos, tr_end)
        if sz is None:
            pos = _resync(slab, pos + 1, tr_end, tr_start, res)
            if pos < 0:
                break
            continue
        nxt = -1
        for cand in (pos + sz, pos + sz + 8):
            if cand <= pos:
                continue
            if cand == tr_end or cand >= tr_end:
                nxt = tr_end
                break
            if _read_header(slab, cand, tr_end) is not None:
                nxt = cand
                break
        if nxt < 0:
            r = _resync(slab, pos + 8, tr_end, tr_start, res)
            nxt = r if r > pos else tr_end
        blocks.append((pos, nxt))
        pos = nxt
    res.meta["blocks"] = len(blocks)

    # Clauses 4 and 5: keep the last block that holds ASCII GPS records.
    chosen = None
    for start, end in blocks:
        lig = start + 8
        if not slab.covers(lig, len(LIGO_MAGIC)):
            continue
        if slab.bytes(lig, len(LIGO_MAGIC)) != LIGO_MAGIC:
            continue
        first = lig + 0x14
        # The record body starts after the 4-byte binary index, so the ASCII test starts there.
        if not is_ascii_region(slab, first + 4, 96):
            continue
        if not is_date_shape(slab, first + 4):
            continue
        chosen = (lig, first, end)
    if chosen is None:
        raise ParseError("trailer has no ASCII LIGOGPSINFO block")
    lig, first, block_end = chosen

    res.meta["variant"] = slab.u8(lig + 0x0B) if slab.covers(lig + 0x0B, 1) else -1
    count_field = slab.u32be(lig + 0x10) if slab.covers(lig + 0x10, 4) else 0
    res.meta["count_field"] = count_field

    # Clause 6.
    stride, warn = detect_stride(slab, first, block_end)
    if warn:
        res.warn(warn)
    if stride <= 0:
        raise ParseError("could not determine record stride")
    res.meta["stride"] = stride

    # Clause 7.
    derived = (block_end - first) // stride
    if derived < 0:
        derived = 0
    count = count_field if count_field > 0 else derived
    if count > derived:
        if count_field > 0:
            res.warn(
                "record count field says %d but only %d records fit; using %d"
                % (count_field, derived, derived)
            )
        count = derived
    elif count_field > 0 and derived - count_field > 1:
        res.warn("record count field says %d but %d records fit" % (count_field, derived))
    res.meta["records"] = count

    # Clause 8.
    tz = opts.tz_offset_s
    dropped = 0
    for i in range(count):
        off = first + i * stride
        idx = slab.u32be(off) if slab.covers(off, 4) else -1
        text = record_text(slab, off, stride)
        if not text:
            continue
        fields = parse_ligo_text(text)
        if fields is None:
            res.warn("unparsable record at index %d" % i)
            continue
        if fields.nofix:
            dropped += 1
            continue
        res.points.append(
            Point(
                fields.t - tz, fields.lat, fields.lon, fields.speed_kmh, fields.heading_deg,
                fields.alt_m, fields.magvar_deg, fields.ax, fields.ay, fields.az, idx, 0,
            )
        )
    res.dropped_nofix = dropped
    return res

File: js/src/formats/ligo_ts_trailer.js
// LigoGPS trailer appended after the last TS packet. spec/01-ligogps-ts-trailer.md
// Clause numbers refer to §1.6. Mirrors python/src/dashgps/formats/ligo_ts_trailer.py
//
// Status: VERIFIED against ~1,230 real ICESKY files.

import { ascii, bytesEqual, readSlab } from "../io.js";
import {
  LIGO_MAGIC, detectStride, isAsciiRegion, isDateShape, parseLigoText, recordText,
} from "../ligo_record.js";
import { ParseError, ParseResult, Point } from "../model.js";

export const FORMAT_ID = "ligo.ts_trailer";
export const STATUS = "verified";

const END_MAGICS = [ascii("####"), ascii("&&&&")];
const TAG_UPPER = ascii("SKIP");
const TAG_LOWER = ascii("skip");
const FOOTER_PROBE = 65536; // one read covers a typical whole trailer
const MAX_BLOCKS = 256;

function isTag(b) { return bytesEqual(b, TAG_UPPER) || bytesEqual(b, TAG_LOWER); }

function memo(opts, value) {
  if (opts) opts.probe["ligo.footer"] = value;
  return value;
}

// Clause 1. Memoized on opts.probe: sniff and parse both need the footer, and on a network
// filesystem a second round trip per file is the difference between seconds and minutes.
async function footer(reader, opts) {
  if (opts && Object.prototype.hasOwnProperty.call(opts.probe, "ligo.footer")) {
    return opts.probe["ligo.footer"];
  }
  const n = reader.size();
  if (n < 32) return memo(opts, null);
  const slab = await readSlab(reader, n > FOOTER_PROBE ? n - FOOTER_PROBE : 0, n);
  if (!slab.covers(n - 8, 8)) return memo(opts, null);
  const magic = slab.bytes(n - 8, 4);
  let ok = false;
  for (const m of END_MAGICS) if (bytesEqual(magic, m)) ok = true;
  if (!ok) return memo(opts, null);
  const length = slab.u32be(n - 4);
  if (length < 8 || length > n) return memo(opts, null);
  let magicStr = "";
  for (const b of magic) magicStr += String.fromCharCode(b);
  return memo(opts, { slab, length, magic: magicStr });
}

export async function sniff(reader, opts) {
  return (await footer(reader, opts)) !== null ? 0.95 : 0.0;
}

// Clause 3
function readHeader(slab, off, trEnd) {
  if (off + 8 > trEnd || !slab.covers(off, 8)) return null;
  if (!isTag(slab.bytes(off + 4, 4))) return null;
  return slab.u32be(off);
}

function resync(slab, after, trEnd, trStart, res) {
  let p = after;
  for (;;) {
    const a = slab.find(TAG_UPPER, p, trEnd);
    const b = slab.find(TAG_LOWER, p, trEnd);
    let hit;
    if (a < 0) hit = b;
    else if (b < 0) hit = a;
    else hit = a < b ? a : b;
    if (hit < 0) return -1;
    const start = hit - 4;
    if (start >= slab.base && readHeader(slab, start, trEnd) !== null) {
      const sz = slab.u32be(start);
      if (sz >= 8 && sz <= trEnd - start + 8) {
        // Offset from the start of the trailer: stable regardless of how much of the
        // file the reader happened to buffer.
        res.warn("resynchronised to block header at trailer+" + (start - trStart));
        return start;
      }
    }
    p = hit + 4;
  }
}

export async function parse(reader, opts) {
  const res = new ParseResult(FORMAT_ID, STATUS, [reader.name]);
  const n = reader.size();

  const f = await footer(reader, opts);
  if (f === null) throw new ParseError("no LigoGPS trailer footer");
  res.meta.end_magic = f.magic;
  res.meta.trailer_len = f.length;

  // Clause 2. Reuse the footer probe when the whole trailer already fits inside it.
  const trStart = n - f.length;
  let readFrom = trStart;
  if (f.length > opts.tailCap) {
    readFrom = n - opts.tailCap;
    res.warn("trailer is " + f.length + " bytes, larger than the " + opts.tailCap +
             "-byte cap; parsing the tail only");
  }
  const slab = f.slab.base <= readFrom ? f.slab : await readSlab(reader, readFrom, n);
  const trEnd = n - 8;

  // Clause 3
  const blocks = [];
  let pos = trStart >= slab.base ? trStart : slab.base;
  let guard = 0;
  while (pos + 8 <= trEnd && guard < MAX_BLOCKS) {
    guard += 1;
    const sz = readHeader(slab, pos, trEnd);
    if (sz === null) {
      pos = resync(slab, pos + 1, trEnd, trStart, res);
      if (pos < 0) break;
      continue;
    }
    let nxt = -1;
    for (const cand of [pos + sz, pos + sz + 8]) {
      if (cand <= pos) continue;
      if (cand >= trEnd) { nxt = trEnd; break; }
      if (readHeader(slab, cand, trEnd) !== null) { nxt = cand; break; }
    }
    if (nxt < 0) {
      const r = resync(slab, pos + 8, trEnd, trStart, res);
      nxt = r > pos ? r : trEnd;
    }
    blocks.push([pos, nxt]);
    pos = nxt;
  }
  res.meta.blocks = blocks.length;

  // Clauses 4 and 5: keep the last block that holds ASCII GPS records.
  let chosen = null;
  for (const [start, end] of blocks) {
    const lig = start + 8;
    if (!slab.covers(lig, LIGO_MAGIC.length)) continue;
    if (!bytesEqual(slab.bytes(lig, LIGO_MAGIC.length), LIGO_MAGIC)) continue;
    const first = lig + 0x14;
    // The record body starts after the 4-byte binary index, so the ASCII test starts there.
    if (!isAsciiRegion(slab, first + 4, 96)) continue;
    if (!isDateShape(slab, first + 4)) continue;
    chosen = [lig, first, end];
  }
  if (chosen === null) throw new ParseError("trailer has no ASCII LIGOGPSINFO block");
  const [lig, first, blockEnd] = chosen;

  res.meta.variant = slab.covers(lig + 0x0b, 1) ? slab.u8(lig + 0x0b) : -1;
  const countField = slab.covers(lig + 0x10, 4) ? slab.u32be(lig + 0x10) : 0;
  res.meta.count_field = countField;

  // Clause 6
  const [stride, warn] = detectStride(slab, first, blockEnd);
  if (warn) res.warn(warn);
  if (stride <= 0) throw new ParseError("could not determine record stride");
  res.meta.stride = stride;

  // Clause 7
  let derived = Math.floor((blockEnd - first) / stride);
  if (derived < 0) derived = 0;
  let count = countField > 0 ? countField : derived;
  if (count > derived) {
    if (countField > 0) {
      res.warn("record count field says " + countField + " but only " + derived +
               " records fit; using " + derived);
    }
    count = derived;
  } else if (countField > 0 && derived - countField > 1) {
    res.warn("record count field says " + countField + " but " + derived + " records fit");
  }
  res.meta.records = count;

  // Clause 8
  const tz = opts.tzOffsetS;
  let dropped = 0;
  for (let i = 0; i < count; i++) {
    const off = first + i * stride;
    const idx = slab.covers(off, 4) ? slab.u32be(off) : -1;
    const text = recordText(slab, off, stride);
    if (!text) continue;
    const fields = parseLigoText(text);
    if (fields === null) { res.warn("unparsable record at index " + i); continue; }
    if (fields.nofix) { dropped += 1; continue; }
    res.points.push(new Point(fields.t - tz, fields.lat, fields.lon, fields.speedKmh,
      fields.headingDeg, fields.altM, fields.magvarDeg, fields.ax, fields.ay, fields.az, idx, 0));
  }
  res.droppedNofix = dropped;
  return res;
}
